radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -69,40 +71,105 @@
 	private final Map<Hash, Vertex> vertices = new ConcurrentHashMap<>();
 	private final Map<Hash, SyncState> syncing = new ConcurrentHashMap<>();
 
-	// TODO: Cleanup this interface
 	public VertexStore(
-		Vertex genesisVertex,
+		Vertex rootVertex,
 		QuorumCertificate rootQC,
 		SyncedStateComputer<CommittedAtom> syncedStateComputer,
 		SyncVerticesRPCSender syncVerticesRPCSender,
 		SyncSender syncSender,
 		SystemCounters counters
 	) {
-		this.syncedStateComputer = syncedStateComputer;
-		this.syncVerticesRPCSender = syncVerticesRPCSender;
-		this.syncSender = syncSender;
+		this(rootVertex, rootQC, Collections.emptyList(), syncedStateComputer, syncVerticesRPCSender, syncSender, counters);
+	}
+
+	public VertexStore(
+		Vertex rootVertex,
+		QuorumCertificate rootQC,
+		List<Vertex> vertices,
+		SyncedStateComputer<CommittedAtom> syncedStateComputer,
+		SyncVerticesRPCSender syncVerticesRPCSender,
+		SyncSender syncSender,
+		SystemCounters counters
+	) {
+		this.syncedStateComputer = Objects.requireNonNull(syncedStateComputer);
+		this.syncVerticesRPCSender = Objects.requireNonNull(syncVerticesRPCSender);
+		this.syncSender = Objects.requireNonNull(syncSender);
 		this.counters = Objects.requireNonNull(counters);
-		this.highestQC.set(Objects.requireNonNull(rootQC));
+
+		Objects.requireNonNull(rootVertex);
+		Objects.requireNonNull(rootQC);
+		Objects.requireNonNull(vertices);
+
+		this.rebuild(rootVertex, rootQC, vertices);
+	}
+
+	private void rebuild(Vertex rootVertex, QuorumCertificate rootQC, List<Vertex> vertices) {
+		if (!rootQC.getProposed().getId().equals(rootVertex.getId())) {
+			throw new IllegalStateException(String.format("rootQC=%s does not match rootVertex=%s", rootQC, rootVertex));
+		}
+
+		this.vertices.clear();
 		this.highestCommittedQC.set(rootQC);
+		this.highestQC.set(rootQC);
+		this.rootId.set(rootVertex.getId());
+		this.vertices.put(rootVertex.getId(), rootVertex);
 
-		if (genesisVertex.getAtom() != null) {
-			CommittedAtom committedGenesis = genesisVertex.getAtom().committed(rootQC.getProposed());
-			syncedStateComputer.execute(committedGenesis);
+		for (Vertex vertex : vertices) {
+			if (!addQC(vertex.getQC())) {
+				throw new IllegalStateException(String.format("Missing qc=%s", vertex.getQC()));
+			}
+
+			try {
+				insertVertexInternal(vertex);
+			} catch (VertexInsertionException e) {
+				throw new IllegalStateException("Could not insert vertex " + vertex);
+			}
 		}
+	}
 
-		this.rootId.set(genesisVertex.getId());
-		this.vertices.put(genesisVertex.getId(), genesisVertex);
-		this.lastCommittedVertex.onNext(genesisVertex);
+	private enum SyncStage {
+		PREPARING,
+		GET_COMMITTED_VERTICES,
+		SYNC_TO_COMMIT, GET_QC_VERTICES
```

Comment:
should be on a new line according to our style guide
```suggestion
		SYNC_TO_COMMIT,
		GET_QC_VERTICES
```